python/tvm/micro/debugger.py

```diff
@@ -82,23 +73,56 @@ def _stop_all(cls):
     def __init__(self):
         super(GdbDebugger, self).__init__()
         self._is_running = False
-        self._child_alive_lock = threading.RLock()
-        self._is_child_alive = False
+        self._is_running_lock = threading.RLock()
+        self._child_exited_event = threading.Event()
+        self._signals_reset_event = threading.Event()
 
     @abc.abstractmethod
     def popen_kwargs(self):
         raise NotImplementedError()
 
+    def _internal_stop(self):
+        if not self._is_running:
+            return
+
+        os.kill(os.getpid(), signal.SIGUSR1)
+        self._signals_reset_event.wait()
+        termios.tcsetattr(sys.stdin.fileno(), termios.TCSAFLUSH, self.old_termios)
+
+        try:
+            children = psutil.Process(self.popen.pid).children(recursive=True)
+            for c in children:
+                c.terminate()
+                _, alive = psutil.wait_procs(children, timeout=self._GRACEFUL_SHUTDOWN_TIMEOUT_SEC)
+                for a in alive:
+                    a.kill()
+        except psutil.NoSuchProcess:
+            pass
+        finally:
+            self.__class__._STARTED_INSTANCE = None
+            self._is_running = False
+            self._child_exited_event.set()
+
     def _wait_for_child(self):
         self.popen.wait()
-        with self._child_alive_lock:
-            self._is_child_alive = True
+        with self._is_running_lock:
+            self._internal_stop()
+
+    @classmethod
+    def _sigusr1_handler(cls, signum, stack_frame):  # pylint: disable=unused-argument
+        if cls._STARTED_INSTANCE is not None:
```

Comment:
Just curious, why not raise an error when STARTED_INSTANCE is None?

Reply:
good call, i've inverted the logic and made the check stricter.

Reply:
the logic seems not inverted yet. missed something in the submit?

Reply:
oooops good catch. actually fixed that now.